consensus/event_loop.go

```diff
@@ -415,14 +417,135 @@ func (fc *FastHotStuffEventLoop) resetScheduledTasks() {
 	}
 
 	// Schedule the next block construction task. This will run with currentView param.
-	fc.nextBlockConstructionTask.Schedule(fc.blockConstructionInterval, fc.currentView, fc.onBlockConstructionScheduledTask)
+	fc.nextBlockConstructionTask.Schedule(fc.blockConstructionInterval, fc.currentView, fc.onBlockConstructionScheduledTaskExecuted)
 
 	// Schedule the next timeout task. This will run with currentView param.
 	fc.nextTimeoutTask.Schedule(timeoutDuration, fc.currentView, fc.onTimeoutScheduledTaskExecuted)
 }
 
-func (fc *FastHotStuffEventLoop) onBlockConstructionScheduledTask(blockConstructionView uint64) {
-	// TODO
+// When this function is triggered, it means that we have reached the block construction
+// time ETA for blockConstructionView. If we have a QC or timeout QC for the view, then we
+// signal the server.
+func (fc *FastHotStuffEventLoop) onBlockConstructionScheduledTaskExecuted(blockConstructionView uint64) {
+	fc.lock.Lock()
+	defer fc.lock.Unlock()
+
+	// Check if the consensus instance is running. If it's not running, then there's nothing
+	// to do here.
+	if fc.status != consensusStatusRunning {
+		return
+	}
+
+	// Check for race conditions where the view advanced at the exact moment this task began
+	// or we have already signaled for this view. If so, then there's nothing to do here.
+	if fc.currentView != blockConstructionView {
+		return
+	}
+
+	// Check if the conditions are met to construct a QC from votes for the chain tip. If so,
+	// we send a signal to the server and cancel the block construction task. The server will
+	// reschedule the task when it advances the view.
+	if canConstructQC, signersList, signature := fc.tryConstructVoteQCInCurrentView(); canConstructQC {
+		// Signal the server that we can construct a QC for the chain tip
+		fc.ConsensusEvents <- &ConsensusEvent{
+			EventType:   ConsensusEventTypeConstructVoteQC, // The event type
+			View:        fc.currentView,                    // The current view in which we can construct a block
+			BlockHash:   fc.tip.block.GetBlockHash(),       // Block hash for the tip, which we are extending from
```

Comment:
What do you think about renaming:
-  BlockHash -> TipBlockHash
-  BlockHeight -> TipBlockHeight (whilst moving the +1 to the receivers)

This will disambiguate that we're referring to the tip and not the would-be block.

Reply:
Hmmm, that works really well and is less ambiguous for all event types (vote, QC construciton, timeout). I've renamed both fields to your suggestions. Thanks!